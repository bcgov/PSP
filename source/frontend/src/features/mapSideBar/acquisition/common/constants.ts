export const cannotEditMessage =
  'This section of the file is in a non-editable state. Contact a System administrator to change it to an editable state';
