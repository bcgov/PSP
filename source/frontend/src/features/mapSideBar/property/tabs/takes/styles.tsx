import { Section } from 'components/common/Section/Section';
import styled from 'styled-components';

export const StyledNoTabSection = styled(Section)`
  &.form-section {
    margin-left: 0;
    margin-right: 0;
    padding-left: 0;
    padding-right: 0;
  }
`;

export const StyledBorderSection = styled(Section)`
  border: 0.1rem solid lightgrey;
  border-radius: 0.5rem;
`;
