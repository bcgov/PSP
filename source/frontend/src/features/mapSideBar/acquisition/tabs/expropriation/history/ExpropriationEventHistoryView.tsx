import React from 'react';
import { FaPlus } from 'react-icons/fa';

import { Section } from '@/components/common/Section/Section';
import { SimpleSectionHeader } from '@/components/common/SimpleSectionHeader';
import { StyledSectionAddButton } from '@/components/common/styles';
import { TableSort } from '@/components/Table/TableSort';

import { ExpropriationEventResults } from './list/ExpropriationEventResults';
import { ExpropriationEventRow } from './models';

export interface IExpropriationEventHistoryViewProps {
  isLoading?: boolean;
  eventRows: ExpropriationEventRow[];
  sort: TableSort<ExpropriationEventRow>;
  setSort: (value: TableSort<ExpropriationEventRow>) => void;
  onAdd: () => void;
  onUpdate: (expropriationEventId: number) => void;
  onDelete: (expropriationEventId: number) => void;
}

export const ExpropriationEventHistoryView: React.FunctionComponent<
  IExpropriationEventHistoryViewProps
> = ({ isLoading, eventRows, sort, setSort, onAdd, onUpdate, onDelete }) => {
  return (
    <Section
      header={
        <SimpleSectionHeader title="Expropriation Date History">
          <StyledSectionAddButton data-testid="add-expropriation-event" onClick={onAdd}>
            <FaPlus size="2rem" />
            &nbsp;{'Add'}
          </StyledSectionAddButton>
        </SimpleSectionHeader>
      }
      isCollapsable
      initiallyExpanded={false}
    >
      <ExpropriationEventResults
        loading={isLoading}
        results={eventRows}
        onUpdate={onUpdate}
        onDelete={onDelete}
        sort={sort}
        setSort={setSort}
      />
    </Section>
  );
};

export default ExpropriationEventHistoryView;
