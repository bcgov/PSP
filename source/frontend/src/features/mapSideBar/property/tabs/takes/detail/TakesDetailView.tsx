import { Col, Row } from 'react-bootstrap';
import { FaPlus, FaTrash } from 'react-icons/fa';
import styled from 'styled-components';

import { StyledRemoveLinkButton } from '@/components/common/buttons';
import EditButton from '@/components/common/EditButton';
import LoadingBackdrop from '@/components/common/LoadingBackdrop';
import { Section } from '@/components/common/Section/Section';
import { SectionField } from '@/components/common/Section/SectionField';
import { StyledEditWrapper, StyledSummarySection } from '@/components/common/Section/SectionStyles';
import { SectionListHeader } from '@/components/common/SectionListHeader';
import { H2 } from '@/components/common/styles';
import TooltipIcon from '@/components/common/TooltipIcon';
import AreaContainer from '@/components/measurements/AreaContainer';
import { Claims, Roles } from '@/constants';
import * as API from '@/constants/API';
import { isAcquisitionFile } from '@/features/mapSideBar/acquisition/add/models';
import StatusUpdateSolver from '@/features/mapSideBar/acquisition/tabs/fileDetails/detail/statusUpdateSolver';
import useKeycloakWrapper from '@/hooks/useKeycloakWrapper';
import useLookupCodeHelpers from '@/hooks/useLookupCodeHelpers';
import { getDeleteModalProps, useModalContext } from '@/hooks/useModalContext';
import { ApiGen_CodeTypes_AcquisitionTakeStatusTypes } from '@/models/api/generated/ApiGen_CodeTypes_AcquisitionTakeStatusTypes';
import { ApiGen_CodeTypes_LandActTypes } from '@/models/api/generated/ApiGen_CodeTypes_LandActTypes';
import { ApiGen_Concepts_FileProperty } from '@/models/api/generated/ApiGen_Concepts_FileProperty';
import { ApiGen_Concepts_Take } from '@/models/api/generated/ApiGen_Concepts_Take';
import { getApiPropertyName, prettyFormatDate, prettyFormatUTCDate } from '@/utils';
import { booleanToYesNoUnknownString } from '@/utils/formUtils';

import { StyledBorderSection, StyledNoTabSection } from '../styles';

export interface ITakesDetailViewProps {
  takes: ApiGen_Concepts_Take[];
  allTakesCount: number;
  loading: boolean;
  fileProperty: ApiGen_Concepts_FileProperty;
  onEdit: (takeId: number) => void;
  onAdd: () => void;
  onDelete: (takeId: number) => void;
}

export const TakesDetailView: React.FunctionComponent<ITakesDetailViewProps> = ({
  takes,
  allTakesCount,
  fileProperty,
  loading,
  onEdit,
  onAdd,
  onDelete,
}) => {
  const cancelledTakes = takes?.filter(
    t => t.takeStatusTypeCode?.id === ApiGen_CodeTypes_AcquisitionTakeStatusTypes.CANCELLED,
  );
  const nonCancelledTakes = takes?.filter(
    t => t.takeStatusTypeCode?.id !== ApiGen_CodeTypes_AcquisitionTakeStatusTypes.CANCELLED,
  );
  const takesNotInFile = allTakesCount - (takes?.length ?? 0);

  const { getCodeById } = useLookupCodeHelpers();
  const { hasClaim, hasRole } = useKeycloakWrapper();
  const { setModalContent, setDisplayModal } = useModalContext();

  const file = fileProperty.file;

  const statusSolver = new StatusUpdateSolver(isAcquisitionFile(file) ? file : null);

  const canEditTakes = (take: ApiGen_Concepts_Take) => {
    if (
      statusSolver.canEditTakes() &&
      ((take.takeStatusTypeCode.id !==
        ApiGen_CodeTypes_AcquisitionTakeStatusTypes.COMPLETE.toString() &&
        hasClaim(Claims.ACQUISITION_EDIT)) ||
        hasRole(Roles.SYSTEM_ADMINISTRATOR))
    ) {
      return true;
    }
    return false;
  };

  return (
    <StyledSummarySection>
      <LoadingBackdrop show={loading} parentScreen={true} />
      <Section>
        <H2>Takes for {getApiPropertyName(fileProperty.property).value}</H2>
        <StyledBlueSection>
          <SectionField
            labelWidth="8"
            label="Takes for this property in the current file"
            tooltip="The number of takes in completed, In-progress or cancelled state for this property in this acquisition file"
          >
            {takes?.length ?? 0}
          </SectionField>
          <SectionField
            labelWidth="8"
            label="Takes for this property in other files"
            tooltip="The number of takes in completed, In-progress or cancelled state for this property, in files other than this acquisition file. The other files can be found under the Acquisition section of the PIMS Files tab"
            valueTestId="takes-in-other-files"
          >
            {takesNotInFile}
          </SectionField>
        </StyledBlueSection>
      </Section>
      <Section>
        <H2 className="mb-8">
          <SectionListHeader
            title="Takes"
            claims={[Claims.PROPERTY_EDIT]}
            addButtonIcon={<FaPlus />}
            addButtonText="Add Take"
            onAdd={statusSolver.canEditTakes() ? onAdd : undefined}
          />
        </H2>
        {[...nonCancelledTakes, ...cancelledTakes].map((take, index) => {
          return (
            <Section
              noPadding
              isCollapsable
              initiallyExpanded
              data-testid={`take-${index}`}
              key={`takes-${index}`}
              header={
                <Row>
                  <Col md="10">Take {index + 1}</Col>
                  <Col md="2" className="d-flex align-items-center justify-content-end">
                    <StyledEditWrapper>
                      {onEdit !== undefined && canEditTakes(take) ? (
                        <EditButton title="Edit take" onClick={() => onEdit(take.id)} />
                      ) : null}
                      {!canEditTakes(take) && (
                        <TooltipIcon
                          toolTipId={`${fileProperty?.fileId || 0}-summary-cannot-edit-tooltip`}
                          toolTip="Retired records are referenced for historical purposes only and cannot be edited or deleted. If the take has been added in error, contact your system administrator to re-open the file, which will allow take deletion"
                        />
                      )}
                    </StyledEditWrapper>
                    {canEditTakes(take) && (
                      <StyledRemoveLinkButton
                        title="Remove take"
                        variant="light"
                        onClick={() => {
                          setModalContent({
                            ...getDeleteModalProps(),
                            handleOk: () => {
                              onDelete(take.id);
                              setDisplayModal(false);
                            },
                          });
                          setDisplayModal(true);
                        }}
                      >
                        <FaTrash size="2rem" />
                      </StyledRemoveLinkButton>
                    )}
                  </Col>
                </Row>
              }
            >
              <SectionField label="Take added on">
                {prettyFormatUTCDate(take.appCreateTimestamp)}
              </SectionField>
              <SectionField label="Take type *">
                {take.takeTypeCode?.id ? getCodeById(API.TAKE_TYPES, take.takeTypeCode.id) : ''}
              </SectionField>
              <SectionField label="Take status *">
                {take.takeStatusTypeCode?.id
                  ? getCodeById(API.TAKE_STATUS_TYPES, take.takeStatusTypeCode.id)
                  : ''}
              </SectionField>
              {take.completionDt && (
                <SectionField label="Completion date *">
                  {prettyFormatDate(take.completionDt)}
                </SectionField>
              )}
              <SectionField label="Site contamination">
                {take.takeSiteContamTypeCode?.id
                  ? getCodeById(API.TAKE_SITE_CONTAM_TYPES, take.takeSiteContamTypeCode.id)
                  : ''}
              </SectionField>
              <SectionField label="Description" labelWidth="12">
                {take.description}
              </SectionField>
              <StyledNoTabSection header="Area">
                <StyledBorderSection>
                  <SectionField
                    label="Is there a new highway dedication? *"
                    labelWidth="9"
                    tooltip="The term new highway dedication includes municipal road or provincial public highway"
                  >
                    {booleanToYesNoUnknownString(take.isNewHighwayDedication ?? undefined)}
                  </SectionField>
                  {take.isNewHighwayDedication && (
                    <SectionField label="Area" labelWidth="12">
                      <AreaContainer landArea={take.newHighwayDedicationArea ?? undefined} />
                    </SectionField>
                  )}
                  <SectionField
                    label="Is this being acquired for MoTI inventory? *"
                    labelWidth="9"
                    tooltip="The property will be added to inventory"
                    className="pt-4"
                  >
                    {booleanToYesNoUnknownString(take.isAcquiredForInventory ?? undefined)}
                  </SectionField>
                </StyledBorderSection>
                <StyledBorderSection>
                  <SectionField
                    label="Is there a new registered interest in land (SRW, Easement or Covenant)? *"
                    labelWidth="9"
                  >
                    {booleanToYesNoUnknownString(take.isNewInterestInSrw ?? undefined)}
                  </SectionField>
                  {take.isNewInterestInSrw && (
                    <>
                      <SectionField label="Area" labelWidth="12">
                        <AreaContainer landArea={take.statutoryRightOfWayArea ?? undefined} />
                      </SectionField>

                      <SectionField label="SRW end date" labelWidth="3" contentWidth="4">
                        {prettyFormatDate(take.srwEndDt ?? undefined)}
                      </SectionField>
                    </>
                  )}
                </StyledBorderSection>
                <StyledBorderSection>
                  <SectionField label="Is there a new Land Act tenure? *" labelWidth="9">
                    {booleanToYesNoUnknownString(take.isNewLandAct ?? undefined)}
                  </SectionField>
                  {take.isNewLandAct && (
                    <>
                      <SectionField label="Land Act" labelWidth="3">
                        {take.landActTypeCode
                          ? take.landActTypeCode.id + ' ' + take.landActTypeCode.description
                          : ''}
                      </SectionField>

                      <SectionField label="Area" labelWidth="12">
                        <AreaContainer landArea={take.landActArea ?? undefined} />
                      </SectionField>

                      {![
                        ApiGen_CodeTypes_LandActTypes.TRANSFER_OF_ADMIN_AND_CONTROL.toString(),
                        ApiGen_CodeTypes_LandActTypes.CROWN_GRANT.toString(),
                      ].includes(take.landActTypeCode.id) && (
                        <SectionField label="End date" labelWidth="3" contentWidth="4">
                          {prettyFormatDate(take.landActEndDt ?? undefined)}
                        </SectionField>
                      )}
                    </>
                  )}
                </StyledBorderSection>
                <StyledBorderSection>
                  <SectionField
                    label="Is there a new Licence for Construction Access (TLCA/LTC)? *"
                    labelWidth="9"
                  >
                    {booleanToYesNoUnknownString(take.isNewLicenseToConstruct ?? undefined)}
                  </SectionField>
                  {take.isNewLicenseToConstruct && (
                    <>
                      <SectionField label="Area" labelWidth="12">
                        <AreaContainer landArea={take.licenseToConstructArea ?? undefined} />
                      </SectionField>

                      <SectionField label="LTC end date" labelWidth="3" contentWidth="4">
                        {prettyFormatDate(take.ltcEndDt ?? undefined)}
                      </SectionField>
                    </>
                  )}
                </StyledBorderSection>
                <StyledBorderSection>
                  <SectionField label="Is there a Lease (Payable)? *" labelWidth="9">
                    {booleanToYesNoUnknownString(take.isLeasePayable ?? undefined)}
                  </SectionField>
                  {take.isLeasePayable && (
                    <>
                      <SectionField label="Area" labelWidth="12">
                        <AreaContainer landArea={take.leasePayableArea ?? undefined} />
                      </SectionField>

                      <SectionField label="End date" labelWidth="3" contentWidth="4">
                        {prettyFormatDate(take.leasePayableEndDt ?? undefined)}
                      </SectionField>
                    </>
                  )}
                </StyledBorderSection>
              </StyledNoTabSection>
              <StyledNoTabSection header="Surplus">
                <StyledBorderSection>
                  <SectionField label="Is there a Surplus? *" labelWidth="9">
                    {booleanToYesNoUnknownString(take.isThereSurplus ?? undefined)}
                  </SectionField>
                  {take.isThereSurplus && (
                    <SectionField label="Area" labelWidth="12">
                      <AreaContainer landArea={take.surplusArea ?? undefined} />
                    </SectionField>
                  )}
                </StyledBorderSection>
              </StyledNoTabSection>
            </Section>
          );
        })}
      </Section>
    </StyledSummarySection>
  );
};

const StyledBlueSection = styled.div`
  background-color: ${({ theme }) => theme.css.filterBoxColor};
  border-radius: 0.5rem;
  padding: 1rem;
`;

export default TakesDetailView;
