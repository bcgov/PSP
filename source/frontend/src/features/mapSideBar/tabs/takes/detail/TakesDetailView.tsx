import YesNoButtons from 'components/common/buttons/YesNoButtons';
import EditButton from 'components/common/EditButton';
import { H2, StyledDivider } from 'components/common/styles';
import LoadingBackdrop from 'components/maps/leaflet/LoadingBackdrop/LoadingBackdrop';
import AreaContainer from 'components/measurements/AreaContainer';
import * as API from 'constants/API';
import { Claims } from 'constants/claims';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import useLookupCodeHelpers from 'hooks/useLookupCodeHelpers';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { Api_Take } from 'models/api/Take';
import * as React from 'react';
import styled from 'styled-components';
import { formatApiDateTime, prettyFormatDate } from 'utils';
import { getApiPropertyName } from 'utils/mapPropertyUtils';

import { Section } from '../../Section';
import { SectionField } from '../../SectionField';
import {
  StyledBorderSection,
  StyledEditWrapper,
  StyledNoTabSection,
  StyledSummarySection,
} from '../styles';

export interface ITakesDetailViewProps {
  takes: Api_Take[];
  allTakesCount: number;
  loading: boolean;
  fileProperty: Api_PropertyFile;
  onEdit: (edit: boolean) => void;
}

export const TakesDetailView: React.FunctionComponent<ITakesDetailViewProps> = ({
  takes,
  allTakesCount,
  fileProperty,
  loading,
  onEdit,
}) => {
  const cancelledTakes = takes?.filter(t => t.takeStatusTypeCode === 'CANCELLED');
  const nonCancelledTakes = takes?.filter(t => t.takeStatusTypeCode !== 'CANCELLED');

  const { getCodeById } = useLookupCodeHelpers();
  const { hasClaim } = useKeycloakWrapper();
  return (
    <StyledSummarySection>
      <LoadingBackdrop show={loading} parentScreen={true} />
      <StyledEditWrapper>
        {onEdit !== undefined && hasClaim(Claims.PROPERTY_EDIT) && (
          <EditButton
            title="Edit takes"
            onClick={() => {
              onEdit(true);
            }}
          />
        )}
      </StyledEditWrapper>
      <Section>
        <H2>Takes for {getApiPropertyName(fileProperty.property).value}</H2>
        <StyledBlueSection>
          <SectionField labelWidth="6" label="Total takes for this property">
            {allTakesCount ?? 0}
          </SectionField>
          <i>Count is inclusive of all files</i>
          <SectionField
            labelWidth="6"
            label="Total cancelled takes for this property"
            className="mt-3"
          >
            {cancelledTakes?.length ?? 0}
          </SectionField>
          <i>Count is inclusive of all files</i>
          <StyledDivider />
          <p>
            There are <b>{takes?.length ?? 0} take(s)</b> for this property on this file.
          </p>
        </StyledBlueSection>
      </Section>
      {[...nonCancelledTakes, ...cancelledTakes].map((take, index) => (
        <Section key={take.id} isCollapsable initiallyExpanded data-testid={`take-${index}`}>
          <H2>Take {index + 1}</H2>
          <SectionField label="Take added on">
            {prettyFormatDate(formatApiDateTime(take.appCreateTimestamp))}
          </SectionField>
          <SectionField label="Take type *">
            {take.takeTypeCode ? getCodeById(API.TAKE_TYPES, take.takeTypeCode) : ''}
          </SectionField>
          <SectionField label="Take status *">
            {take.takeStatusTypeCode
              ? getCodeById(API.TAKE_STATUS_TYPES, take.takeStatusTypeCode)
              : ''}
          </SectionField>
          <SectionField label="Site contamination">
            {take.takeSiteContamTypeCode
              ? getCodeById(API.TAKE_SITE_CONTAM_TYPES, take.takeSiteContamTypeCode)
              : ''}
          </SectionField>
          <SectionField label="Description" labelWidth="12">
            {take.description}
          </SectionField>
          <StyledNoTabSection header="Area">
            <StyledBorderSection>
              <SectionField label="Is there a new right of way? *" labelWidth="8">
                <YesNoButtons id="newRightOfWayToggle" disabled value={take.isNewRightOfWay} />
              </SectionField>
              {take.isNewRightOfWay && (
                <SectionField label="Area" labelWidth="12">
                  <AreaContainer landArea={take.newRightOfWayArea ?? undefined} />
                </SectionField>
              )}
            </StyledBorderSection>
            <StyledBorderSection>
              <SectionField label="Is there a Statutory Right of Way: (SRW)? *" labelWidth="8">
                <YesNoButtons
                  id="statutoryRightOfWayToggle"
                  disabled
                  value={take.isStatutoryRightOfWay}
                />
              </SectionField>
              {take.isStatutoryRightOfWay && (
                <>
                  <SectionField label="Area" labelWidth="12">
                    <AreaContainer landArea={take.statutoryRightOfWayArea ?? undefined} />
                  </SectionField>
                </>
              )}
            </StyledBorderSection>
            <StyledBorderSection>
              <SectionField label="Is there a Section 16? *" labelWidth="8">
                <YesNoButtons id="landActToggle" disabled value={take.isLandAct} />
              </SectionField>
              {take.isLandAct && (
                <>
                  <SectionField label="Area" labelWidth="12">
                    <AreaContainer landArea={take.landActArea ?? undefined} />
                  </SectionField>
                  {take.landActEndDt && (
                    <SectionField label="Section 16 end date">
                      {prettyFormatDate(take.landActEndDt ?? undefined)}
                    </SectionField>
                  )}
                </>
              )}
            </StyledBorderSection>
            <StyledBorderSection>
              <SectionField label="Is there a License to Construct (LTC)? *" labelWidth="8">
                <YesNoButtons
                  id="licenseToConstructToggle"
                  disabled
                  value={take.isLicenseToConstruct}
                />
              </SectionField>
              {take.isLicenseToConstruct && (
                <>
                  <SectionField label="Area" labelWidth="12">
                    <AreaContainer landArea={take.licenseToConstructArea ?? undefined} />
                  </SectionField>
                  {take.ltcEndDt && (
                    <SectionField label="LTC end date">
                      {prettyFormatDate(take.ltcEndDt ?? undefined)}
                    </SectionField>
                  )}
                </>
              )}
            </StyledBorderSection>
          </StyledNoTabSection>
          <StyledNoTabSection header="Surplus">
            <StyledBorderSection>
              <SectionField label="Is there a Surplus? *" labelWidth="8">
                <YesNoButtons id="surplusToggle" disabled value={take.isSurplus} />
              </SectionField>
              {take.isSurplus && (
                <SectionField label="Area" labelWidth="12">
                  <AreaContainer landArea={take.surplusArea ?? undefined} />
                </SectionField>
              )}
            </StyledBorderSection>
          </StyledNoTabSection>
        </Section>
      ))}
    </StyledSummarySection>
  );
};

const StyledBlueSection = styled.div`
  background-color: ${({ theme }) => theme.css.filterBoxColor};
  border-radius: 0.5rem;
  padding: 0.5rem;
`;

export default TakesDetailView;
