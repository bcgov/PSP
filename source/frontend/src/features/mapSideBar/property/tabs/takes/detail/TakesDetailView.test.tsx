import { createMemoryHistory } from 'history';

import { Claims } from '@/constants/claims';
import { mockLookups } from '@/mocks/lookups.mock';
import { getMockApiPropertyFiles } from '@/mocks/properties.mock';
import { getMockApiTakes } from '@/mocks/takes.mock';
import { ApiGen_Concepts_File } from '@/models/api/generated/ApiGen_Concepts_File';
import { lookupCodesSlice } from '@/store/slices/lookupCodes';
import { toTypeCodeNullable } from '@/utils/formUtils';
import { act, render, RenderOptions, screen, userEvent, within } from '@/utils/test-utils';

import TakesDetailView, { ITakesDetailViewProps } from './TakesDetailView';
import { ApiGen_CodeTypes_AcquisitionStatusTypes } from '@/models/api/generated/ApiGen_CodeTypes_AcquisitionStatusTypes';

const history = createMemoryHistory();
const storeState = {
  [lookupCodesSlice.name]: { lookupCodes: mockLookups },
};

const onEdit = jest.fn();
jest.mock('@react-keycloak/web');

describe('TakesDetailView component', () => {
  // render component under test
  const setup = (renderOptions: RenderOptions & { props?: Partial<ITakesDetailViewProps> }) => {
    const utils = render(
      <TakesDetailView
        {...renderOptions.props}
        takes={renderOptions.props?.takes ?? []}
        allTakesCount={renderOptions.props?.allTakesCount ?? 0}
        loading={renderOptions.props?.loading ?? false}
        fileProperty={renderOptions.props?.fileProperty ?? getMockApiPropertyFiles()[0]}
        onEdit={onEdit}
      />,
      {
        ...renderOptions,
        claims: renderOptions.claims ?? [],
        store: storeState,
        history,
      },
    );

    return {
      ...utils,
    };
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', () => {
    const { asFragment } = setup({});
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays a loading spinner when loading', () => {
    const { getByTestId } = setup({ props: { loading: true } });
    const spinner = getByTestId('filter-backdrop-loading');
    expect(spinner).toBeVisible();
  });

  it('clicking the edit button fires the edit event', async () => {
    const { getByTitle } = setup({ props: { loading: true }, claims: [Claims.PROPERTY_EDIT] });
    const editButton = getByTitle('Edit takes');
    await act(async () => userEvent.click(editButton));
    expect(onEdit).toHaveBeenCalled();
  });

  it('hides the edit button when the file has been completed', () => {
    const fileProperty = getMockApiPropertyFiles()[0];
    const file: ApiGen_Concepts_File = fileProperty!.file as ApiGen_Concepts_File;
    const { queryByTitle, getByTestId } = setup({
      props: {
        loading: true,
        fileProperty: {
          ...fileProperty,
          file: {
            ...file,
            fileStatusTypeCode: toTypeCodeNullable(ApiGen_CodeTypes_AcquisitionStatusTypes.COMPLT),
          },
        },
      },
      claims: [Claims.PROPERTY_EDIT],
    });
    const editButton = queryByTitle('Edit takes');
    expect(editButton).toBeNull();
    const tooltip = getByTestId('tooltip-icon-1-summary-cannot-edit-tooltip');
    expect(tooltip).toBeVisible();
  });

  it('displays the number of takes in other files', () => {
    setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            takeStatusTypeCode: toTypeCodeNullable('CANCELLED'),
            id: 1,
          },
          { ...getMockApiTakes()[0], takeStatusTypeCode: toTypeCodeNullable('INPROGRESS') },
        ],
        allTakesCount: 10,
      },
    });
    const takesNotInFile = screen.getByTestId('takes-in-other-files');
    expect(takesNotInFile).toHaveTextContent('8');
  });

  it('displays all non-cancelled takes and then cancelled takes', () => {
    setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            takeStatusTypeCode: toTypeCodeNullable('CANCELLED'),
            id: 1,
          },
          { ...getMockApiTakes()[0], takeStatusTypeCode: toTypeCodeNullable('INPROGRESS') },
        ],
      },
    });
    const { getByText } = within(screen.getByTestId('take-0'));
    expect(getByText('In-progress')).toBeVisible();
  });

  it('does not display an area fields if all is radio buttons are false', () => {
    const { queryAllByText } = setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            isNewLicenseToConstruct: false,
            isNewHighwayDedication: false,
            isNewLandAct: false,
            isNewInterestInSrw: false,
            isThereSurplus: false,
            isLeasePayable: false,
          },
        ],
      },
    });
    expect(queryAllByText('Area:')).toHaveLength(0);
  });

  it('displays all area fields if all is radio buttons are true', () => {
    const { getAllByText } = setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            isNewLicenseToConstruct: true,
            isNewHighwayDedication: true,
            isNewLandAct: true,
            isNewInterestInSrw: true,
            isThereSurplus: true,
            isLeasePayable: true,
          },
        ],
      },
    });
    expect(getAllByText('Area:')).toHaveLength(6);
  });

  it('displays srwEndDt if specified', async () => {
    const { findByText } = setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            isNewInterestInSrw: true,
            srwEndDt: '2020-01-01',
          },
        ],
      },
    });
    const date = await findByText('SRW end date:');
    expect(date).toBeVisible();
  });

  it('displays ltcEndDt if specified', async () => {
    const { findByText } = setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            isNewLicenseToConstruct: true,
            ltcEndDt: '2020-01-01',
          },
        ],
      },
    });
    const date = await findByText('LTC end date:');
    expect(date).toBeVisible();
  });

  it('displays landActEndDt if specified', async () => {
    const { findByText } = setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            isNewLandAct: true,
            landActEndDt: '2020-01-01',
          },
        ],
      },
    });
    const date = await findByText('Is there a new Land Act tenure', {
      exact: false,
    });
    expect(date).toBeVisible();
  });

  it('displays leasePayableEndDt if specified', async () => {
    const { findByText } = setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            isNewLandAct: true,
            leasePayableEndDt: '2022-11-21',
          },
        ],
      },
    });
    const date = await findByText('Is there a Lease (Payable)', {
      exact: false,
    });
    expect(date).toBeVisible();
  });

  it('displays the land act code', async () => {
    const { findByText } = setup({
      props: {
        loading: false,
        takes: [
          {
            ...getMockApiTakes()[0],
            isNewLicenseToConstruct: true,
            ltcEndDt: '2020-01-01',
          },
        ],
      },
    });
    const code = await findByText('Section 15', {
      exact: false,
    });
    expect(code).toBeVisible();
  });
});
