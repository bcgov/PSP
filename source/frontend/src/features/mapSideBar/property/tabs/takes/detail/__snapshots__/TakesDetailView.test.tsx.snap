// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TakesDetailView component > renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  .c2 {
  font-family: 'BCSans-Bold';
  font-size: 2.6rem;
  border-bottom: solid 0.2rem;
  width: 100%;
  text-align: left;
  margin-bottom: 2.4rem;
}

.c0 {
  padding-top: 1rem;
}

.c1 {
  margin: 1.6rem;
  padding: 1.6rem;
  background-color: white;
  text-align: left;
  border-radius: 0.5rem;
}

.c5.required::before {
  content: '*';
  position: absolute;
  top: 0.75rem;
  left: 0rem;
}

.c4 {
  font-weight: bold;
}

.c6 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  min-height: 4.5rem;
}

.c6 .btn {
  margin: 0;
}

.c3 {
  border-radius: 0.5rem;
  padding: 1rem;
}

<div
    class="c0"
  >
    <div
      class="c1 form-section"
    >
      <div
        class="collapse show"
      >
        <h2
          class="c2"
        >
          Takes for 007-723-385
        </h2>
        <div
          class="c3"
        >
          <div
            class="pb-2 row"
          >
            <div
              class="pr-0 text-left col-8"
            >
              <label
                class="c4"
              >
                Takes for this property in the current file:
                <span
                  class="ml-2"
                >
                  <span
                    class="tooltip-icon"
                    data-testid="tooltip-icon-section-field-tooltip"
                    id="section-field-tooltip"
                  >
                    <svg
                      class="tooltip-icon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                      />
                    </svg>
                  </span>
                </span>
              </label>
            </div>
            <div
              class="c5 text-left col"
            >
              0
            </div>
          </div>
          <div
            class="pb-2 row"
          >
            <div
              class="pr-0 text-left col-8"
            >
              <label
                class="c4"
              >
                Takes for this property in other files:
                <span
                  class="ml-2"
                >
                  <span
                    class="tooltip-icon"
                    data-testid="tooltip-icon-section-field-tooltip"
                    id="section-field-tooltip"
                  >
                    <svg
                      class="tooltip-icon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                      />
                    </svg>
                  </span>
                </span>
              </label>
            </div>
            <div
              class="c5 text-left col"
              data-testid="takes-in-other-files"
            >
              0
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c1 form-section"
    >
      <div
        class="collapse show"
      >
        <h2
          class="c2 mb-8"
        >
          <div
            class="c6 no-gutters row"
          >
            <div
              class="px-2 my-1 col-auto"
            >
              Takes
            </div>
            <div
              class="my-1 col-auto"
            />
          </div>
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
